### PR TITLE
Uploads backend

### PR DESCRIPTION
**Pull request by katch00**
capture (2)

Sorry it's so long :( I couldn't wait for all the PR's to be approved so I just copy pasted my code into this branch based off of Indhu's attribution branch. I need to create a new form field so the user can add alt text, but will do that in a later PR. The screen capture is of the deployed site and I'm merging the deployed link.

### CODE REVIEW

File: backend/src/main/java/com/google/capstone/ServeUploads.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that gets all stored entities from Datastore
+ * Turns entities to UplaodedImage objects and converts to json
```
Comment (indhu-gunda) [inline backend/src/main/java/com/google/capstone/ServeUploads.java:26]: nit: UploadedImage

File: backend/src/main/java/com/google/capstone/ServeUploads.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * then returns the json in response
+ */
+@WebServlet("/api/v1/serveUploads")
+public class ServeUploads extends HttpServlet {
```
Comment (indhu-gunda) [inline backend/src/main/java/com/google/capstone/ServeUploads.java:30]: Maybe rename to getUploads to reduce confusion with serving like ServeBlob does?

File: backend/src/main/java/com/google/capstone/UserUploadHandler.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+ * Servlet used to handle image upload to Google Cloud Storage using the Blobstore API
+ * Also uploads Image information, blobkey, and servingURL to the Datastore
+ * Note: must be updated before deployment to redirect to proper frontend server
+ */
+@WebServlet("/api/v1/uploadImgs")
+public class UserUploadHandler extends HttpServlet{
+ 
+  public static final String GCS_BUCKET_NAME = "upload-imgs";
+ 
+  private final BlobstoreService blobstore = BlobstoreServiceFactory.getBlobstoreService();
```
Comment (indhu-gunda) [inline backend/src/main/java/com/google/capstone/UserUploadHandler.java:34]: I think these can be static too.

File: frontend/src/components/dashboard/UserUploadForm.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+                type="submit"
+              >
+                Submit Artwork
+              </Button>
+            </div>
+          </Grid>
+        </form>
+      </>
+  );
+}
```
Comment (indhu-gunda) [inline frontend/src/components/dashboard/UserUploadForm.js:165]: add new line at end of file

File: frontend/src/components/dashboard/UserUploadForm.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+     }
+  };
+
+  useEffect(() => {
+    fetch('/api/v1/uploadImgs')
+        .then((response) => {
+          return response.text();
+        })
+        .then((imageUploadUrl) => {
+          console.log(imageUploadUrl);
```
Comment (isabelsang) [inline frontend/src/components/dashboard/UserUploadForm.js:47]: nit: remove

File: backend/src/main/java/com/google/capstone/ServeUploads.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Servlet that gets all stored entities from Datastore
+ * Turns entities to UploadedImage objects and converts to json
+ * then returns the json in response
+ */
+@WebServlet("/api/v1/serveUploads")
+public class ServeUploads extends HttpServlet {
```
Comment (BearBurg) [inline backend/src/main/java/com/google/capstone/ServeUploads.java:30]: This class seems to be a dupe of UploadInformation?

File: backend/src/main/java/com/google/capstone/UserUploadHandler.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+      mssgEntity.setProperty("timestamp", timestamp);
+ 
+      String blobKey = blobKeys.get(0).getKeyString();
+      mssgEntity.setProperty("blobKey", blobKey);
+      mssgEntity.setProperty("url", "/api/v1/get-blob?blob-key=" + blobKey);
+      
+      datastore.put(mssgEntity);  
+      
+      //deployment link
+      response.sendRedirect("https://igunda-isangimino-nstroupe.uc.r.appspot.com/user-uploads-gallery");
```
Comment (BearBurg) [inline backend/src/main/java/com/google/capstone/UserUploadHandler.java:91]: Using absolute path doesn't seems right here. Check https://www.codejava.net/java-ee/servlet/how-to-send-redirect-from-java-servlet for how to use relative path. If you still have problem with redirecting, I am happy to debug with you.

File: backend/src/main/java/com/google/capstone/UserUploadHandler.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+      //ImageInformation entity to be added into datastore
+      Entity mssgEntity = new Entity("ImageInformation");
+      mssgEntity.setProperty("artistName", artistName);
+      mssgEntity.setProperty("artTitle", artTitle);
+      mssgEntity.setProperty("description", description);
+      mssgEntity.setProperty("timestamp", timestamp);
+ 
+      String blobKey = blobKeys.get(0).getKeyString();
+      mssgEntity.setProperty("blobKey", blobKey);
+      mssgEntity.setProperty("url", "/api/v1/get-blob?blob-key=" + blobKey);
```
Comment (BearBurg) [inline backend/src/main/java/com/google/capstone/UserUploadHandler.java:86]: The `url` seems to be redundant here. We should be able to build the `url` by looking at the `blobKey`.

File: backend/src/main/java/com/google/capstone/UserUploadHandler.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  @Override
+  protected void doPost(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+      Map<String, List<BlobKey>> blobs = blobstore.getUploads(request);
+      List<BlobKey> blobKeys = blobs.get("selectedFile"); 
+ 
+      //checks to make sure image was added to create blobkey
+      if (blobKeys == null || blobKeys.isEmpty()) {
+        /* NOTE: add error page here, no file selected!!
+           Will create better error handling in future update :) 
+           Currently redirect to museumGallery for testing reasons */
```
Comment (BearBurg) [inline backend/src/main/java/com/google/capstone/UserUploadHandler.java:66]: nit: use // for comments in code.

File: frontend/src/redux/userArtworkActions.js [hunk 10]
```diff
@@ -1,98 +1,10 @@
...
-}
-
 function convertToArtworkInfo(artwork) {
   const artworkInfo = new Map();
   artworkInfo.set('id', artwork.id);
-  artworkInfo.set('alt', artwork.alt);
-  artworkInfo.set('url', artwork.img);
-  artworkInfo.set('title', artwork.title);
-  artworkInfo.set('artist', artwork.artist);
+  //artworkInfo.set('alt', artwork.alt);
```
Comment (BearBurg) [inline frontend/src/redux/userArtworkActions.js:4]: remove unused line?

File: backend/src/main/java/com/google/capstone/ServeUploads.java (hunk 6 above)
Comment (katch00) [reply backend/src/main/java/com/google/capstone/ServeUploads.java:30]: UploadInformation was deleted, this is renamed and has a slightly different functionality

**Pull request merged**